File: Cpp/Beginner/03_BinarySearch/findPart.cpp
//
// Created by seungsu on 22. 11. 18.
//

#include <iostream>
#include <algorithm>
#include <unordered_set>

using namespace std;

constexpr int None = -1;

static int binary_search_recursive(int* arr, int start, int end, int part) {
    if (start > end)
        return None;

    int mid = (start + end) / 2;

    if (arr[mid] == part)
        return mid;
    else if (arr[mid] > part)
        return binary_search_recursive(arr, start, mid - 1, part);
    else
        return binary_search_recursive(arr, mid + 1, end, part);
}

// 이진 탐색을 이용한 풀이
void find_part_test1() {
    int n;
    cin >> n;

    int* arr = new int[n];
    for (int i = 0; i < n; i++)
        cin >> arr[i];

    std::sort(arr, arr + n);

    int m;
    cin >> m;

    for (int i = 0; i < m; i++) {
        int part;
        cin >> part;

        int result = binary_search_recursive(arr, 0, n - 1, part);

        if (result == None) cout << "no" << "\n";
        else cout << "yes" << "\n";
    }

    delete[] arr;
}

// unordered_set 자료 구조를 이용한 풀이
void find_part_test2() {
    int n;
    cin >> n;

    unordered_set<int> s;

    for (int i = 0; i < n; i++) {
        int x;
        cin >> x;
        s.insert(x);
    }

    int m;
    cin >> m;

    for (int i = 0; i < m; i++) {
        int part;
        cin >> part;

        // 찾지 못한 경우 end() 반환
        if (s.find(part) == s.end()) cout << "no" << "\n";
        else cout << "yes" << "\n";
    }
}

File: Python/beginner/03_BinarySearch/findPart.py
import sys

input = sys.stdin.readline

# 손님이 요청한 m개의 부품이 있으면 yes를, 없으면 no를 출력하자!

# 이진탐색을 이용한 풀이
def bin_search(arr, start, end, target):
    if start > end:
        return None

    mid = (start + end) // 2
    
    if arr[mid] == target:
        return mid
    elif arr[mid] > target:
        return bin_search(arr, start, mid - 1, target)
    else:
        return bin_search(arr, mid + 1, end, target)

if __name__ == '__main__':
    n = int(input())
    arr = list(map(int, input().split()))
    arr.sort()

    m = int(input())
    find = list(map(int, input().split()))

    for f in find:
        result = bin_search(arr, 0, n - 1, f)
        print("no") if result == None else print("yes")
        
# 그 외에도 계수정렬 개념을 이용한 풀이나, 
# 집합 자료형을 이용한 풀이가 가능하다 
# if f in set -> yes, no
